src/Kernel/RectorKernel.php

```diff
@@ -15,51 +15,75 @@
 use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\DependencyInjection\Dumper\PhpDumper;
+use Symplify\SmartFileSystem\SmartFileSystem;
+use Webmozart\Assert\Assert;
 
 final class RectorKernel
 {
     private readonly ConfigureCallValuesCollector $configureCallValuesCollector;
 
     private ContainerInterface|null $container = null;
 
+    /**
+     * @var array<string, ContainerBuilder>
+     */
+    private static array $kernelsByHash = [];
+
     public function __construct()
     {
         $this->configureCallValuesCollector = new ConfigureCallValuesCollector();
     }
 
     /**
      * @api used in tests
+     *
+     * @param string[] $configFiles
      */
-    public function create(): ContainerInterface
+    public function createBuilder(array $configFiles = []): ContainerBuilder
     {
-        return $this->createFromConfigs([]);
+        return $this->buildContainer($configFiles);
     }
 
     /**
+     * @api used in tests
+     *
      * @param string[] $configFiles
      */
-    public function createFromConfigs(array $configFiles): ContainerBuilder
+    public function createFromConfigs(array $configFiles): ContainerInterface
     {
-        $defaultConfigFiles = $this->createDefaultConfigFiles();
-        $configFiles = array_merge($defaultConfigFiles, $configFiles);
-
-        $compilerPasses = $this->createCompilerPasses();
-
-        $configureCallMergingLoaderFactory = new ConfigureCallMergingLoaderFactory($this->configureCallValuesCollector);
-        $containerBuilderFactory = new ContainerBuilderFactory($configureCallMergingLoaderFactory);
-
-        $containerBuilder = $containerBuilderFactory->create($configFiles, $compilerPasses);
-
-        // @see https://symfony.com/blog/new-in-symfony-4-4-dependency-injection-improvements-part-1
-        $containerBuilder->setParameter('container.dumper.inline_factories', true);
-        // to fix reincluding files again
-        $containerBuilder->setParameter('container.dumper.inline_class_loader', false);
+        if ($configFiles === []) {
+            return $this->buildContainer([]);
+        }
 
-        $containerBuilder->compile();
+        $hash = $this->createConfigsHash($configFiles);
+        if (isset(self::$kernelsByHash[$hash])) {
+            return self::$kernelsByHash[$hash];
+        }
 
-        $this->container = $containerBuilder;
+        $className = 'RectorKernel'.$hash;
+        $file = sys_get_temp_dir() .'/rector/kernel-v3-'.$hash.'.php';
```

Comment:
I prefer it to pull from config, there is container cache directory for that:

https://github.com/rectorphp/rector-src/blob/7da67eebfc9011b316e08d80be83a6223f83a371/config/config.php#LL75C29-L75C29

it can be pulled by:

```php
$parameterProvider->provideStringParameter(Option::CONTAINER_CACHE_DIRECTORY)
```

if that pullable from this area.

Reply:
see the possible bug that will showing again on php 8.x with direct non-configurable `sys_get_temp_dir()` on run rector:

-  https://github.com/rectorphp/rector/issues/7700

Reply:
I don't know, probably better to make it optional if possible, eg something like `--dump-container` ?

Reply:
we only need this feature in the test-suite. do you have an idea on how to enable it for test-suite runs ony?

Reply:
Check phpunit constant may can do that

https://github.com/rectorphp/rector-src/blob/7da67eebfc9011b316e08d80be83a6223f83a371/packages/Testing/PHPUnit/StaticPHPUnitEnvironment.php#L12-L15

Reply:
fixed

Reply:
ha, had alredy forgotten about it :).

fixed